File: matrix.cpp
// Demonstration of representing/analyzing circuits as MNA matrices
// to accompany "Analyzing On-Chip Interconnect with Modern C++"
// Author: Jeff Trull <edaskel@att.net>

/*
Copyright (c) 2014 Jeffrey E. Trull

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

#include <iostream>
#include <boost/numeric/odeint.hpp>

#include "ckt_matrix.h"

// Create my standard 2-signal coupling testcase
struct coupling_circuit_t {
    coupling_circuit_t() {
        // MNA
        // MNA - we will have 10 state variables:
        // 8 for node voltages (vagg, n1, n2, n3, vvic, n5, n6, n7)
        // 2 for input currents (iagg, ivic)
        using namespace Eigen;
        typedef Matrix<double, 10, 10> state_matrix_t;
        state_matrix_t G, C;   // conductance and time derivative matrices
        G = state_matrix_t::Zero();
        C = state_matrix_t::Zero();

        double const kohm = 1000;
        double const ff   = 1e-15;
        double rdrv       = 0.1*kohm;
        double pi_r       = 1.0*kohm;
        double pi_c       = 100*ff;
        double coupl_c    = 100*ff;
        double rcvr_c     = 20*ff;

        stamp_i(G, 0, 8);                // aggressor driver current
        stamp(G, 0, 1, 1.0 / rdrv);      // aggressor driver impedance
        stamp(C, 1,    pi_c / 2);        // begin first "pi"
        stamp(G, 1, 2, 1.0 / pi_r);
        stamp(C, 2,    pi_c / 2);        // central node
        stamp(C, 2,    pi_c / 2);        // second "pi"
        stamp(G, 2, 3, 1.0 / pi_r);
        stamp(C, 3,    pi_c / 2);
        stamp(C, 3,    rcvr_c);          // aggressor receiver

        stamp_i(G, 4, 9);                // victim driver current
        stamp(G, 4, 5, 1.0 / rdrv);      // victim driver impedance
        stamp(C, 5,    pi_c / 2);        // begin first "pi"
        stamp(G, 5, 6, 1.0 / pi_r);
        stamp(C, 6,    pi_c / 2);        // central node
        stamp(C, 6,    pi_c / 2);        // second "pi"
        stamp(G, 6, 7, 1.0 / pi_r);
        stamp(C, 7,    pi_c / 2);
        stamp(C, 7,    rcvr_c);          // victim receiver

        stamp(C, 2, 6, coupl_c);         // coupling cap

        // We also have 2 inputs, two outputs
        typedef Matrix<double, 10, 2> io_matrix_t;
        io_matrix_t B = io_matrix_t::Zero();
        B(8, 0) = -1;                    // connect input 0 to vagg
        B(9, 1) = -1;                    // connect input 1 to vvic
        io_matrix_t L = io_matrix_t::Zero();
        L(3, 0) = 1;                     // connect agg rcvr to output 0
        L(7, 1) = 1;                     // connect vic rcvr to output 1

        // cross-check: compute moments
        Matrix<double, 2, 2> E = Matrix<double, 2, 2>::Zero();   // feedthrough term we don't have
        auto block_moments = moments(G, C, B, L, E, 2);
        std::cerr << "moment 0=\n" << block_moments[0] << std::endl;
        std::cerr << "moment 1=\n" << block_moments[1] << std::endl;

        // Now regularize.  Results are of dynamic (not initially known) size
        MatrixXd Greg, Creg;             // regularized versions of C and G
        Matrix<double, Dynamic, 2> Breg, Lreg;
        std::tie(Greg, Creg, Breg, Lreg) = regularize(G, C, B, L);

        // Finally, put in a form suitable for simulation, by transforming
        // C*dX/dt = -G*X + B*u
        // into
        // dX/dt   = -C.inv()*G*X + C.inv()*B*u
        // Wikipedia says Cholesky decomposition "roughly twice as efficient" as LU:
        assert(canLDLTDecompose(Creg));            // make sure we can use it
        drift_  = Creg.ldlt().solve(-1.0 * Greg);  // -Creg.inv()*Greg
        input_  = Creg.ldlt().solve(Breg);         // Creg.inv()*Breg
        output_ = Lreg.transpose();

    }

    // perform dX/dt calculation for ODEInt
    typedef std::vector<double> state_t;
    void operator()(state_t const& x, state_t& dxdt, double) const {
        using namespace Eigen;
        // need to wrap std::vector state types for Eigen to use
        Map<const Matrix<double, Dynamic, 1>> xvec(x.data(), x.size());
        Map<Matrix<double, Dynamic, 1>>       result(dxdt.data(), x.size());

        // simulating step function at time 0 for simplicity
        Matrix<double, 2, 1> u; u << 1.0, 0.0;   // aggressor voltage 1V, victim quiescent
        result = drift_ * xvec + input_ * u;
    }
        
    // turns internal state into output by applying transformed L matrix
    std::vector<double> state2output(state_t const& x) const {
        using namespace Eigen;
        std::vector<double> result(2);
        Map<const Matrix<double, Dynamic, 1> > xvec(x.data(), x.size());
        Map<Matrix<double, 2, 1> >             ovec(result.data());
        ovec = output_ * xvec;
        return result;
    }

    size_t statecnt() const {
        return drift_.rows();
    }

private:
    Eigen::Matrix<double, Eigen::Dynamic, 2>              input_;
    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> drift_;
    Eigen::Matrix<double, 2, Eigen::Dynamic>             output_;
};

int main() {
    using namespace std;

    // instantiate circuit
    coupling_circuit_t coupling_test;

    // simulate
    typedef coupling_circuit_t::state_t state_t;
    state_t         x(coupling_test.statecnt(), 0.0);   // initial conditions = all zero

    using boost::numeric::odeint::integrate;

    integrate( coupling_test, x, 0.0, 1e-9, 1e-12,
               [&](state_t const& x, double t) {
                   auto outputs = coupling_test.state2output(x);
                   cout << t << " " << outputs[0] << " " << outputs[1] << endl;
               });
}

File: matrix.py
#!/usr/bin/python
# Demonstration of representing/analyzing circuits as MNA matrices
# to accompany "Analyzing On-Chip Interconnect with Modern C++"
# but using Python!
# Author: Jeff Trull <edaskel@att.net>

# Copyright (c) 2014 Jeffrey E. Trull
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

import sys
import numpy as np
import scipy as sp
import scipy.linalg
import scipy.integrate

# functions to construct MNA matrices

# stamp conductance between "i" and "j" nodes
# (resistor -> G, or capacitor -> C)
def stamp(m, i, *args):
    # args can be either:
    if len(args) == 1:
        # for a lumped-to-ground conductance
        m[i, i] += args[0]
    else:
        # OR node2, conductance
        j = args[0]
        g = args[1]
        m[i, i] += g
        m[j, j] += g
        m[i, j] -= g
        m[j, i] -= g

# stamp a voltage source connection into G matrix
def stamp_i(m, v, i):
    m[v, i] = 1
    m[i, v] = -1

# Analysis functions
def isSingular(m):
    if m.shape[0] != m.shape[1]:
        raise ValueError('Singularity test only works on square matrices')

    return np.linalg.matrix_rank(m) != m.shape[0]  # not "full rank"

def moments(G, C, B, L, N):
    # calculate the first N moments of the system
    if isSingular(G):
        raise ValueError('Supplied G matrix is singular, cannot invert')

    A = np.linalg.solve(-G, C)   # runs G decomposition twice - can be optimized
    R = np.linalg.solve(G, B)

    return [np.matrix.transpose(L).dot(np.linalg.matrix_power(A, i)).dot(R)
            for i in xrange(0, N)]

def regularize(G, C, B, L):

    # 1. Generate permutation matrix to move zero rows to the bottom
    zero_rows = np.all(C == 0, axis=1)

    P  = np.identity(10)
    j  = 9
    i  = 0
    while i < j:
        # search for an exchange point
        while i < 10 and not zero_rows[i]:
            i += 1
        while (j > 0) and zero_rows[j]:
            j -= 1
        if i < j:
            P[:, [i, j]] = P[:, [j, i]]  # exchange columns
            i += 1
            j -= 1
            
    # 2. Apply permutation to MNA matrices
    CP = P.dot(C).dot(P)     # P * C * P
    GP = P.dot(G).dot(P)
    BP = P.dot(B)
    LP = P.dot(L)

    # 3. Produce reduced equations following Su (Proc. 15th ASP-DAC, 2002)
    nonzero_count = sum(1 for r in zero_rows if not r)
    G11 = GP[0:nonzero_count, 0:nonzero_count]   # upper left
    G12 = GP[0:nonzero_count, nonzero_count:10]  # upper right
    G21 = GP[nonzero_count:10, 0:nonzero_count]  # upper left
    G22 = GP[nonzero_count:10, nonzero_count:10] # upper left

    L1  = LP[0:nonzero_count]
    L2  = LP[nonzero_count:10]

    B1  = BP[0:nonzero_count]
    B2  = BP[nonzero_count:10]

    Creg = CP[0:nonzero_count, 0:nonzero_count]

    if isSingular(G22):
        raise ValueError("Cannot invert G22")
    g22lu     = sp.linalg.lu_factor(G22)        # create LU object for reuse
    G22invG21 = sp.linalg.lu_solve(g22lu, G21)
    G22invB2  = sp.linalg.lu_solve(g22lu, B2)
    Greg      = G11 - G12.dot(G22invG21)

    Lreg      = (L1.transpose() - L2.transpose().dot(G22invG21)).transpose()
    Breg      = B1 - G12.dot(G22invB2)

    return (Greg, Creg, Breg, Lreg)

def func(y, t0):
    # derivative is linear to produce a parabolic output
    return [t0]

class mna_ckt_fun(object):
    # class to compute dX/dt given X and t
    # that's C^-1*(-G) + C^-1*B*u
    def __init__(self, G, C, B):
        # precompute C^-1*(-G) and C^-1*B
        C_lu = sp.linalg.lu_factor(C)
        self.drift  = sp.linalg.lu_solve(C_lu, -G)
        self.inTerm = sp.linalg.lu_solve(C_lu, B)

    def __call__(self, x, t):
        # return dx/dt given x and t
        u = [1.0, 0]    # vAgg step at time zero; vVic constant 0
        return self.drift.dot(x) + self.inTerm.dot(u)

if __name__ == "__main__":
    # construct C*dX/dt = -G * X + B * u via MNA

    G = np.zeros(shape=(10, 10), dtype=float)
    C = np.zeros(shape=(10, 10), dtype=float)

    kohm       = 1000.0
    ff         = 1e-15
    rdrv       = 0.1*kohm
    pi_r       = 1.0*kohm
    pi_c       = 100*ff
    coupl_c    = 100*ff
    rcvr_c     = 20*ff

    stamp_i(G, 0, 8)                # aggressor driver current
    stamp(G, 0, 1, 1.0 / rdrv)      # aggressor driver impedance
    stamp(C, 1,    pi_c / 2)        # begin first "pi"
    stamp(G, 1, 2, 1.0 / pi_r)
    stamp(C, 2,    pi_c / 2)        # central node
    stamp(C, 2,    pi_c / 2)        # second "pi"
    stamp(G, 2, 3, 1.0 / pi_r)
    stamp(C, 3,    pi_c / 2)
    stamp(C, 3,    rcvr_c)          # aggressor receiver

    stamp_i(G, 4, 9)                # victim driver current
    stamp(G, 4, 5, 1.0 / rdrv)      # victim driver impedance
    stamp(C, 5,    pi_c / 2)        # begin first "pi"
    stamp(G, 5, 6, 1.0 / pi_r)
    stamp(C, 6,    pi_c / 2)        # central node
    stamp(C, 6,    pi_c / 2)        # second "pi"
    stamp(G, 6, 7, 1.0 / pi_r)
    stamp(C, 7,    pi_c / 2)
    stamp(C, 7,    rcvr_c)          # victim receiver

    stamp(C, 2, 6, coupl_c)         # coupling cap

    # finally, construct B (voltage input to state variables)
    B = np.zeros(shape=(10, 2), dtype=float)
    B[8, 0] = -1
    B[9, 1] = -1

    # and L to extract outputs (via Y = L * X)
    L = np.zeros(shape=(10, 2), dtype=float)
    L[3, 0] = 1
    L[7, 1] = 1
    
    block_moments = moments(G, C, B, L, 2)
    print >> sys.stderr, "moment 0="
    print >> sys.stderr, str(block_moments[0])
    print >> sys.stderr, "moment 1="
    print >> sys.stderr, str(block_moments[1])

    (Greg, Creg, Breg, Lreg) = regularize(G, C, B, L)

    ckt_fun = mna_ckt_fun(Greg, Creg, Breg)

    timepoints = np.arange(0, 1e-9, 1e-12)
    xarr = sp.integrate.odeint(ckt_fun,                     # dX/dt functor
                               [0, 0, 0, 0, 0, 0],          # initial state
                               timepoints)

    for (t, x) in zip(timepoints, xarr):
        y = Lreg.transpose().dot(x)
        print t, y[0], y[1]
